protocol/contracts/core/OtcOffer.sol

```diff
@@ -0,0 +1,168 @@
+
+import { CommonEventsAndErrors } from "contracts/common/CommonEventsAndErrors.sol";
+
+contract OtcOffer is Pausable, Ownable {
+    using SafeERC20 for IERC20Metadata;
+
+    /// @notice The token that the user will sell
+    IERC20Metadata public immutable userSellToken;
+
+    /// @notice The token that the user will buy
+    IERC20Metadata public immutable userBuyToken;
+
+    /// @notice Where to pull the `userBuyToken` from, and to send the `userSellToken`
+    /// @dev The funds owner must grant approval for this contract to pull the `userBuyToken`
+    address public fundsOwner;
+    
+    /// @notice The number of decimal places represented by `offerPrice`
+    uint8 public constant OFFER_PRICE_DECIMALS = 18;
+
+    /// @notice The offer price, specified in terms of `offerPricingToken`.
+    uint256 public offerPrice;
+
+    enum OfferPricingToken {
+        /// @notice The offerPrice 'pricing' token is defined in terms of the `userBuyToken`
+        /// ie price is userBuyToken / userSellToken
+        /// eg when user sells OHM to buy DAI, the price is defined in terms of DAI/OHM
+        UserBuyToken,
+
+        /// @notice The offerPrice 'pricing' token is defined in terms of the `userBuyToken`
+        /// ie price is userSellToken / userBuyToken
+        /// eg when user sells DAI to buy OHM, the price can still be defined in terms of DAI/OHM
+        UserSellToken
+    }
+
+    /// @notice Which token the `offerPrice` is defined in terms of.
+    OfferPricingToken public immutable offerPricingToken;
+
+    // @notice How to scale the fixed point buyTokenAmount given differences in token decimal places.
+    uint256 public immutable scalar;
+
+    /// @notice Any update to the `offerPrice` must be within this percentage (as basis points. 20% = 2_000) 
+    /// of the existing `offerPrice`.
+    uint256 public offerPriceUpdateThresholdBps;
+
+    /// @notice 100% represented as basis points.
+    uint256 public constant BPS_100_PCT = 10_000;
+
+    event OfferPriceSet(uint256 _offerPrice);
+    event OfferPriceUpdateThresholdBpsSet(uint256 offerPriceUpdateThresholdBps);
+    event Swap(address indexed account, address indexed fundsOwner, uint256 userSellTokenAmount, uint256 userBuyTokenAmount);
+    event FundsOwnerSet(address indexed fundsOwner);
+
+    error OfferPriceDeltaTooBig(uint256 newOfferPrice, uint256 maxPriceAllowed);
+
+    constructor(
+        address _userSellToken,
+        address _userBuyToken,
+        address _fundsOwner,
+        uint256 _offerPrice,
+        OfferPricingToken _offerPricingToken,
+        uint256 _offerPriceUpdateThresholdBps
+    ) {
+        userSellToken = IERC20Metadata(_userSellToken);
+        userBuyToken = IERC20Metadata(_userBuyToken);
+        fundsOwner = _fundsOwner;
+
+        offerPrice = _offerPrice;
+        offerPricingToken = _offerPricingToken;
+        offerPriceUpdateThresholdBps = _offerPriceUpdateThresholdBps;
+
+        // The price is always specified in 18dp
```

Comment:
It might be better to have the price expressed in the decimals of the pricing token (now that we have a nominated pricing token).

Thoughts?

These details wouldn't matter too much if there was going to be a UI for updating the price. But given that it will happen in terms of etherscan or safe tx builder, we want to minimise the risk of errors.

Reply:
In my opinion given its unlikely to have a UI, it's better to have the price decimals consistent and constant

Reply:
Since it needs to be mindless for lux and pengu

Reply:
Non 18dp is the exception rather than the rule. So my view is that a fixed 18dp seems right